pkg/controllers/leaderworkerset_controller.go

```diff
@@ -246,40 +260,32 @@ func (r *LeaderWorkerSetReconciler) rollingUpdatePartition(ctx context.Context,
 		return 0, err
 	}
 
+	// It can happen that the statefulset is recreating, then the sts list is less than
+	// the length, which also indicates that the missing statefulset is not ready.
+	sortedSts := utils.SortByIndex(func(sts appsv1.StatefulSet) (int, error) {
+		return strconv.Atoi(sts.Labels[leaderworkerset.GroupIndexLabelKey])
+	}, stsList.Items, int(replicas))
+
 	var finishedPart int32
 	templateHash := utils.LeaderWorkerTemplateHash(lws)
 
-stop:
 	for i := replicas - 1; i >= 0; i-- {
-		for j := len(stsList.Items) - 1; j >= 0; j-- {
-			nominatedName := fmt.Sprintf("%s-%d", lws.Name, i)
-			if stsList.Items[j].Name == nominatedName {
-				// TODO: quick sort the pods and statefulsets first, which will be faster than API requests.
-				// Or only construct workerStatefulset until leaderPod is ready.
-
-				// Rolling update happens only when all the former groups of pods(leaderPod+workerStatefulset) are all ready
-				// then we'll go process the next group.
-				// Compare the templateRevisionHash to make sure that we're checking the right object.
-
-				stsTemplateHash := stsList.Items[j].Labels[leaderworkerset.TemplateRevisionHashKey]
-				if !(stsTemplateHash == templateHash && statefulsetutils.StatefulsetReady(stsList.Items[j])) {
-					break stop
-				}
-
-				leaderPod := &corev1.Pod{}
-				if err := r.Get(ctx, types.NamespacedName{Name: nominatedName, Namespace: lws.Namespace}, leaderPod); err != nil {
-					return 0, err
-				}
+		nominatedName := fmt.Sprintf("%s-%d", lws.Name, i)
+		if nominatedName != sortedPods[i].Name || nominatedName != sortedSts[i].Name {
```

Comment:
maybe add a comment here (or move the one above) saying this indicates that the leader or the worker statefulset isn't created yet.